File: Roadmap/04 - CADENAS DE CARACTERES/java/DiegoIBB.java
package Cadena_de_Caracteres_04;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Dictionary;
import java.util.HashMap;
import java.util.Hashtable;


/*
 * EJERCICIO:
 * Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
 * en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
 * - Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
 *   conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que analice dos palabras diferentes y realice comprobaciones
 * para descubrir si son:
 * - Palíndromos
 * - Anagramas
 * - Isogramas
 */


public class Cadena_de_Caracteres {

	public static void main(String[]args) {
		
		//OPERACIONES BASICAS
		
		// Concatenar
		String string_1 = "cadena 1";
		String string_2 = "cadena 2";
		String string_3 = " segundo elemento";

		System.out.println(string_1 + " más " + string_2);
		System.out.println(string_1.concat(string_3)); // Método CONCAT
		
		//Indexar
		String string_4 = "elemento";
		System.out.println(string_4.indexOf("m"));//Por caracter
		
		
		
		// METODOS ESPECIALES
		
		String cadena_1 = "Hola mundo";
		String cadena_2 = "HELLO";
		String cadena_3 = "MINUSCULAS";
		String cadena_4 = "mayusculas";
		String cadena_5 = "Esta es una cadena";
		String cadena_6 = "Caracter";
		String cadena_7 = "Listado de caracteres";
		
		//Método charAt, retorna un carácter a partir de un indice
 		System.out.println("---- METODO CHAR AT ----");
		System.out.println(cadena_1.charAt(3));

		//Método Length cuenta la cantidad de caracteres de una string
 		System.out.println("---- METODO LENGTH ----");
		System.out.println(cadena_1.length());
		
		//Método Replace cuenta la cantidad de caracteres de una string
 		System.out.println("---- METODO REPLACE ----");
		System.out.println(cadena_2.replace("L", "R"));
		
		//Método ToLowerCase convierte los caracteres en mayusculas a minusculas
 		System.out.println("---- METODO LOWERCASE ----");
 		System.out.println(cadena_3.toLowerCase());

		//Método ToUperCase convierte los caracteres en minusculas a mayusculas
 		System.out.println("---- METODO UPPERCASE ----");
 		System.out.println(cadena_4.toUpperCase());
 		
 		//Método Split, convierte una cadena en unas serie de subcadenas en formato array
 		System.out.println("---- METODO SPLIT ----");
 		String [] split_1 = cadena_5.split(" "); //Creamos un array y dividimos la cadena a partir de un caracter espacio(" ")
 		String part_1 = split_1[1]; // Podemos guardar los valores del array en variables
 		
 		System.out.println(split_1[0]);
 		System.out.println(part_1);
 		System.out.println(split_1[2]);
 		System.out.println(split_1[3]);

 		//Metodo endsWith, nos permite saber si una cadena termina con uno o varios caracteres especificos, devuelve true o false
 		System.out.println("---- METODO ENDSWITH ----");
 		System.out.println(cadena_6.endsWith("Cara"));
 		System.out.println(cadena_6.endsWith("ter"));
 		System.out.println(cadena_6.endsWith("r"));
 		
 		//Metodo startsWith, a diferencia de endsWith nos permite saber si una cadena empieza con uno o varios caracteres especificos
 		System.out.println("---- METODO STARTSWITH ----");
 		System.out.println(cadena_6.startsWith("Car"));
 		System.out.println(cadena_6.startsWith("ter"));
 		System.out.println(cadena_6.startsWith("C"));
 		
 		//Metodo getChars, junta un grupo de caracteres en un array 
		char[] ch = new char[7]; //En la declaración del array especificamos la cantidad de caracteres si la string tiene más se muestran espacios en blanco
 		cadena_7.getChars(0, 7, ch, 0); // (caracte-inicial, caracter-final, array, desfase)
 		System.out.println(ch);
 		
 		
		// OPERACIONES
 		
 		//Reversa a una cadena, invertimos el orden de una cadena
 		System.out.println("-------- OPERACION REVERSA -------");
 		String cadenaOriginal = "Cadena al reves";
 		String cadenaReversa = "";
 		
 		for (int i = 0; i < cadenaOriginal.length(); i++) {
 			System.out.println(cadenaOriginal.charAt(i));
 			cadenaReversa = cadenaOriginal.charAt(i) + cadenaReversa;
 		}
 		System.out.println(cadenaReversa);
 		
 		
 		/*
 		* DIFICULTAD EXTRA (opcional):
 		* Crea un programa que analice dos palabras diferentes y realice comprobaciones
 		* para descubrir si son:
 		* - Palíndromos
 		* - Anagramas
 		* - Isogramas
 		*/
 		
 		// ---------- PALINDROMO ---------
 		Scanner palindromo = new Scanner(System.in);
 		System.out.println("Frase:");
 		String palindromo_1 = palindromo.nextLine();
 		String frase_reves = "";
 		String frase_sinEspacios = "";
 		
 		//Pasamos todas las letras a minusculas
 		String palindromo_op = palindromo_1.toLowerCase();
 		
 		//Invierte la frase
 		for (int l=0; l < palindromo_op.length(); l++) {
 			if (palindromo_1.charAt(l) == ' ') {
 				continue;
 			}else if (palindromo_op.charAt(l) != ' ') {
 				frase_reves = palindromo_op.charAt(l) + frase_reves;
 			}
 		}
 		
 		//Frase original sin espacios (invertimos la frase que ya habiamos invertido, esta vez sin espacios)
 		for (int s=0; s < frase_reves.length(); s++){
 			frase_sinEspacios = frase_reves.charAt(s) + frase_sinEspacios;
 		}
 		
 		
 		System.out.println(palindromo_op);
 		System.out.println(frase_sinEspacios);
 		System.out.println(frase_reves);
 		
		if (frase_sinEspacios.equals(frase_reves)) { //CUANDO QUEREMOS COMPARAR EL CONTENIDO DE 2 CADENAS DEBEMOS USAR EL METODO EQUALS()
 			System.out.println("La frase es un palindromo");
 		}else {
 			System.out.println("La frase no es un palindromo ");
 		}
		
		/* Si hacemos una comparación tradicional "if(cadena_1 == cadena_2)"
		 * lo que estamos haciendo es comparar si la instancia es la misma,
		 * sin embargo la instancia no es la misma
		 * 
		 * (instancia) String frase_reves es distinto a (instancia) String palindromo_1 
		 * 
		 * la forma correcta de evaluar si el contenido de 2 cadenas es el mismo es por medio del
		 * método equals()
		 * 
		 * */
				
 		// ---------- ANAGRAMA ---------
				
		//Entregar 2 palabras y ver si una es un anagrama de la otra
		
		//Entrada de palabras
		Scanner palabra_1 = new Scanner(System.in);
		System.out.println("Palabra 1: ");
		String word_1 = palabra_1.nextLine();
		
		Scanner palabra_2 = new Scanner(System.in);
		System.out.println("Palabra 2: ");
		String word_2 = palabra_2.nextLine();
		
		anagrama(word_1, word_2);

 		// ---------- ISOGRAMA ---------
		
		Scanner palabra_3 = new Scanner(System.in);
		System.out.println("Palabra 3: ");
		String word_3 = palabra_3.nextLine();
		
		
		System.out.println(isograma(word_3));
		
		
		Scanner palabra_4 = new Scanner(System.in);
		System.out.println("Palabra 4: ");
		String word_4 = palabra_4.nextLine();
		
		System.out.println(isograma_2(word_4));
		
	}
	
	public static String anagrama(String cadena_1, String cadena_2) {
		
		String cadena_1_toLow;
		String cadena_2_toLow;
		
		//Convertimos las cadenas originales a minusculas
		cadena_1_toLow = cadena_1.toLowerCase(); //Guardamos la primera cadena con todos los caracteres en  minusculas
		cadena_2_toLow = cadena_2.toLowerCase(); //Guardamos la segunda cadena con todos los caracteres en  minusculas

		//Obtenemos el largo de cada cadena
		int cadena_1len = cadena_1.length();
		int cadena_2len = cadena_2.length();
		int counter = 0;
		int charCounter = 0;
		
		ArrayList<Character> letra_cadena_1 = new ArrayList<Character>();
		ArrayList<Character> letra_cadena_2 = new ArrayList<Character>();
		
		for(int c=0; c < cadena_1len; c++) { //Para añadir caracteres de la primera palabra en una lista
			letra_cadena_1.add(cadena_1_toLow.charAt(c));
		}
		for(int d = 0; d < cadena_2len; d++) { //Para añadir caracteres de la segunda palabra en una lista
			letra_cadena_2.add(cadena_2_toLow.charAt(d));
		}
		
		//Recorremos la segunda cadena a partir de cada caracter de la primera cadena
		if(cadena_1len == cadena_2len){
			for(int e = 0; e < cadena_1len; e++) {
				charCounter = 0;
				for(int f = 0; f < cadena_2len; f++) {
					if(cadena_1_toLow.charAt(e) == cadena_2_toLow.charAt(f)) {
						//System.out.println("Element " + cadena_1_toLow.charAt(e) + " is in both strings");
						charCounter += 1;
						counter += 1;
						//Esta sección reduce en 1 el contador si se repite más de una vez //Resolver que no se repita la impresión del mismo caracater 2 veces
						if(charCounter > 1) {
							counter -= 1;
						}
						else {
							System.out.println("Element " + cadena_1_toLow.charAt(e) + " is in both strings");
						}
					}
					else {
						continue;
					}	
				}
				//System.out.println(charCounter);
			}
			
			//Establecemos que 2 palabras son anagramas si tienen los mismos caracteres
			
			if (counter == cadena_1len) {
				System.out.println("The strings are anagrams");
			}else {
				System.out.println("The strings are not anagramas");
			}
		}else {
			System.out.println("The strings are not anagramas");
		}
		
		System.out.println("Length of both strings: " + cadena_1len + " and " + cadena_2len); 
		System.out.println("Characters that coincide: " + cadena_1len + " and " + counter); //Counter difiere del largo de la primera cadena si hay caracteres repetidos

		
		return "La palabra es un anagrama";
	}
	
	
		// ---------- ISOGRAMA ---------

	//Entregar 1 palabra y ver si se repiten la misma cantidad de veces cada una de sus letras
	
	//Señalamos que el resultado a retornar es un Diccionario
	public static Dictionary isograma(String cadena_3) {
		String cadena_3_low;
		cadena_3_low = cadena_3.toLowerCase();
		
		Dictionary<Character, Integer> dictCaracteres = new Hashtable<>();
		
		for(int i=0; i < cadena_3_low.length(); i++) {
				dictCaracteres.put(cadena_3_low.charAt(i), 1);
		}
		
		return dictCaracteres;
	}
	
	
	//Señalamos que el resultado a retornar es un HashMap

	public static HashMap isograma_2(String cadena_3) {
		String cadena_4_low;
		cadena_4_low = cadena_3.toLowerCase();
		
		HashMap<Character, Integer> hashMapCaracteres = new HashMap<>();
		
		for(int i=0; i < cadena_4_low.length(); i++) {
			for(int e=0; e < cadena_4_low.length(); e++) {
				if(cadena_4_low.charAt(i) == hashMapCaracteres.get(e)) {
					hashMapCaracteres.put(cadena_4_low.charAt(i), 1);
				}else {
					hashMapCaracteres.put(cadena_4_low.charAt(i), 2);
				}
			}
			//hashMapCaracteres.put(cadena_4_low.charAt(i), 1);
//			if(cadena_4_low.charAt(i) == hashMapCaracteres.get(cadena_4_low.charAt(i))) { //Buscar método que verifique que un elemento está en el diccionario
//				hashMapCaracteres.put(cadena_4_low.charAt(i), 1);
//			}
		}
		
		return hashMapCaracteres;
	}
	
}

File: Roadmap/04 - CADENAS DE CARACTERES/python/DiegoIBB.py
'''
* EJERCICIO:
 * Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
 * en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
 * - Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
 *   conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que analice dos palabras diferentes y realice comprobaciones
 * para descubrir si son:
 * - Palíndromos
 * - Anagramas
 * - Isogramas
 '''

#-------------------------------------------------
#-------------- OPERACIONES BASICAS --------------
#-------------------------------------------------


#CONCATENAZIÓN ---> Unir 2 cadenas de diferentes variables
string_1 = "La hora es "
string_2 = "21:30"
concatenacion = string_1 + string_2
print(concatenacion)

#INDEXACIÓN ---> Acceder a un caracter especifico de una cadena por medio de su indice
string_3 = "Programacion"
caracter = string_3[5]
print(caracter)




#-------------------------------------------------
#-------------- METODOS CON CADENAS --------------
#-------------------------------------------------


#------ CAPITALIZE ---> Transforma la primera letra de una cadena en mayuscula
cadena_1 = "hola"
c1 = cadena_1.capitalize()
print(c1)

#------ CASEFOLD ---> Convierte una cadena a minusculas
cadena_2 = "HaY MuchAs MayusCuLas"
c2 = cadena_2.casefold()
print(c2)

#------ COUNT ---> Cuenta la cantidad de veces que se repite un elemento de las string
cadena_3 = "pelar manzanas, cocer manzanas, servir manzanas"
c3 = cadena_3.count("manzanas")
print(c3)

#------ FIND ---> Devuelve la cantidad de elementos que coincidan con un caracter a indicar
cadena_4 = "Bienvenido"
c4 = cadena_4.find("i")
print(c4)

#------ JOIN ---> Une todos los elementos de un grupo de elementos con un caracter de conexión
lista_1 = ["Esta", "es", "una", "cadena", "de", "una", "lista"]
c5 = " ".join(lista_1)
print(c5)

#------ PARTITION ---> Divide una cadena en 3 partes y devuelve una tupla
cadena_5 = "las cadenas pueden dividirse con el método partition"
c6 = cadena_5.partition("dividirse")
print(c6)

#------ RSTRIP ---> Nos permite eliminar los espacios vacios al final de una cadena
cadena_6 = "nombre        "
c7 = cadena_6.rstrip()
print("Mi", c7, "es Ignacio")


#------ SPLIT ---> Divide una cadena en sus diferentes palabras y las guarda en una lista
cadena_7 = "las cadenas pueden dividirse usando split"
c8 = cadena_7.split()
print(c8)

#------ RSPLIT ---> Crea una lista a partir de una cadena usando un separador especial
cadena_8 = "Mercedez Benz, Audi, Porsche, BMW, Volkswagen"
c9 = cadena_8.rsplit(", ")
print(c9)
print(c9[3])

#------ TITLE ---> Convierte el primer caracter de cada palabra de una string en mayusculas
cadena_9 = "diego, pablo, felipe, jose, maria, miguel"
c10 = cadena_9.title()
print(c10)

#------ SWAPCASE ---> Convierte mayusculas en minusculas y viceversa
cadena_10 = "Las Jerarquia Se Invierte"
c11 = cadena_10.swapcase()
print(c11)


#------ UPPER ---> Convierte todas las letras a myusculas
cadena_11 = "estoy gritando"
c12 = cadena_11.upper()
print(c12)

#------ REPLACE ---> Reemplaza todos los caracteres que conicidan con uno definido 
cadena_12 = "coincidencia"
c13 = cadena_12.replace("i", "a")
print(c13)


'''
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que analice dos palabras diferentes y realice comprobaciones
 * para descubrir si son:
 * - Palíndromos
 * - Anagramas
 * - Isogramas
 '''


#PALINDROMO: Palabra o frase cuyas letras están dispuestas de tal manera que resulta la misma leída de izquierda a derecha que de derecha a izquierda

cadena_original = input("Palindromo ?: ")
lista_letras = []
lista_2 = [] #Para obtener la cadena original sin espacios
counter = 0

cadena_Reversa = ""
cadena_notReversa = ""
cadena_aux = ""

#Transformar la cadena a lista
for l in cadena_original:
    #print(l)
    lista_letras.append(l)

#Quitar espacios a la cadena original e invertirla
for c in range(0, len(lista_letras)):
    cadena_aux = lista_letras[c]
    if cadena_aux == " ":
        continue
    elif cadena_aux != " ":
        cadena_Reversa = cadena_aux + cadena_Reversa

#Reconstruir la cadena original a partir de la cadena invertida sin espacios
for f in cadena_Reversa:
    lista_2.append(f)
    cadena_notReversa = lista_2[counter] + cadena_notReversa
    counter += 1

print(lista_2) #Lista
print(cadena_notReversa) #Cadena reconstruida sin espacios


print("Cadena Original: " + cadena_notReversa)
print("Cadena Reversa: " + cadena_Reversa)

if cadena_notReversa == cadena_Reversa:
    print("La cadena es un palindromo")
else:
    print("La cadena no es un palindromo")


#ANAGRAMA:

string_org_An = input("Cadena Original: ")
string_comp = input("Anagrama?: ")
counter = 0
charCounter = 0
list_strOrg = []
list_comp = []

for i in string_org_An:
    char1 = i.lower()
    list_strOrg.append(char1)

for h in string_comp:
    char2 = h.lower()
    list_comp.append(char2)

len_cadOrg = len(list_strOrg)
len_cadCom = len(list_comp)

if len_cadOrg == len_cadCom:
    print (len_cadOrg , "=", len_cadCom ) #Evaluamos si el largo de caracteres coincide
    for c in list_strOrg:
        charCounter = 0 #Char counter se actualiza aquí por cada vez que ingresa una nueva letra para ser coparada con las de la segunda cadena
        for d in list_comp:
            if (c == d):
                charCounter+=1
                counter+=1
                if charCounter == 2:
                    counter -= 1
                else:
                    print(c, " is in ", string_comp)
            else:
                continue

    if counter == len_cadOrg:
        print("Strings are anagramas")
    else:
        print("String are not anagramas")
        print(counter, " = ", len_cadOrg)
else:
    print("la cadena no es un anagrama") 


#ISOGRAMA: Palabra o frase donde cada letra aparece el mismo número de veces

isograma = input("Isograma: ")
isogramaToLower = isograma.lower()
lista_iso = []
letra = 0
contador = 0
caracterIsograma = {}

print(isogramaToLower)

for c in isogramaToLower:
    lista_iso.append(c)

for i in lista_iso:
    caracterIsograma[i] = 1
    # letra = i
    # lista_iso.remove(i)
    # for l in lista_iso:
    #     if letra == l:
    #         contador += 1
    #     else:
    #         continue
    # print(f"{letra}, {contador}")

    #caracterIsograma[i] = 1

print(caracterIsograma)

